판매 / 구매 등록 구현 [즉시 구매]

kihyuk-jeong opened the pull request:
1. 즉시 판매 비즈니스 로직 구현
2. 즉시 구매 비즈니스 로직 구현

 #76 판매 / 구매 등록 구현 [즉시 구매]

kihyuk-jeong committed:
[#76] feat : 즉시 구매 / 즉시 판매 비즈니스 로직 구현

1. 즉시 판매 비즈니스 로직 구현
2. 즉시 구매 비즈니스 로직 구현

 #76 판매 / 구매 등록 구현 [즉시 구매]

Hosick commented on src/main/java/com/flab/shoeauction/controller/dto/TradeDto.java:
```diff
@@ -89,4 +90,21 @@ public Trade toEntityByBuyer(User user, Product product, Address address) {
+        private Long tradeId;
+        private Long addressId;
+        private Long productId;
+
```
즉시 거래의 경우 거래를 등록한다기보다 거래를 원하는 상품의 현재 등록된 목록 중 '최저가 + 가장 빠른 등록'의 상품과  거래가 매칭되는 로직이라고 생각되는데 거래 Id를 받아올 필요가 있을까요??

Hosick commented on src/main/java/com/flab/shoeauction/service/TradeService.java:
```diff
@@ -71,4 +74,30 @@ public void createPurchaseBid(String email, SaveRequest requestDto) {
+            .orElseThrow(() -> new UserNotFoundException("존재하지 않는 사용자 입니다."));
+
+        Address shippingAddress = addressRepository.findById(requestDto.getAddressId()).orElseThrow();
+
```
즉시 거래의 경우 구매자와 판매자가 같은 요청이 들어올 수도 있는데 예외처리하는 것이 좋지 않을까요??
(ex. 유저가 A 상품을 판매 등록 후 동일 유저가 A 상품을 즉시 구매 하였을 때 자신이 올린 상품과 매칭되는 경우)

kihyuk-jeong replied:
미처 고려하지 못한 부분이네요. 감사합니다. 예외 처리 로직 추가하겠습니다!

kihyuk-jeong replied:
현재 로그인된 사용자와 Trade의 `publisherId(거래 생성자)`가 동일한 경우, 최고가/최저가 표시에서  제외시켜, 예외 상황 자체가 발생하지 않도록 수정하였습니다!

Hosick commented on src/main/java/com/flab/shoeauction/controller/dto/TradeDto.java:
```diff
@@ -89,4 +90,21 @@ public Trade toEntityByBuyer(User user, Product product, Address address) {
+        @Builder
+        public ImmediateTradeRequest(Long tradeId, Long addressId, Long productId) {
+            this.tradeId = tradeId;
+            this.addressId = addressId;
```
즉시 구매 시 addressId로 주소 정보를 받아온다면 이미 저장된 주소가 아닌 새로운 주소(User와 관계를 갖고 있지 않은 주소)는 사용할 수 없는걸까요?

kihyuk-jeong replied:
네. 이부분은 저도 생각을 했었는데 부가적인 로직이 너무 많이 추가돼서 1차 배포때는 번거롭지만 주소록에 가서 주소를 추가하는 방식으로 진행하려고 합니다 ㅠㅠ

kihyuk-jeong committed:
[#76] refactor : 리뷰 반영

1. 자기 자신이 등록한 거래에 대해 즉시구매/즉시판매 불가능 하도록 수정
2. 거래 등록 페이지에서 필요한 리소스 얻어오는 메소드 네임 수정

kihyuk-jeong committed:
[#76] refactor : 주석 제거

f-lab-dev commented on src/main/java/com/flab/shoeauction/domain/product/Product.java:
```diff
@@ -113,39 +115,47 @@ public ProductInfo toProductInfoResponse() {
     }
 
-    private TradeBidResponse getHighestPrice(double size) {
+    private Predicate<Trade> LowestPriceFilter(User currentUser, double size) {
```
메소드 이름에 카멜케이스를 지켜주세요~

kihyuk-jeong replied:
수정했습니다!

kihyuk-jeong committed:
[#76] refactor : 리뷰 반영

1. 카멜케이스 지키도록 수정
